autopush/router/apnsrouter.py

```diff
@@ -27,86 +28,162 @@ class APNSRouter(object):
             elif "encryption-key" in notification.headers:
-                custom["Enckey"] = notification.headers["encryption-key"]
-
-        payload = apns.Payload(alert=router_data.get("title",
-                                                     self.default_title),
-                               content_available=1,
-                               custom=custom)
-        now = int(time.time())
-        self.messages[now] = {"token": token, "payload": payload}
-        # TODO: Add listener for error handling.
-        self.apns.gateway_server.register_response_listener(self._error)
-        self.ap_settings.metrics.increment(
-            "updates.client.bridge.apns.attempted",
-            self._base_tags)
+                custom["enckey"] = notification.headers["encryption-key"]
 
-        self.apns.gateway_server.send_notification(token, payload, now)
+        payload = apns.Payload(
+            alert=router_data.get("title", config.get('default_title',
+                                                      'Mozilla Push')),
+            content_available=1,
+            custom=custom)
+        now = time.time()
 
-        # cleanup sent messages
-        if self.messages:
-            for time_sent in self.messages.keys():
-                if time_sent < now - self.config.get("expry", 10):
-                    del self.messages[time_sent]
+        self.messages[uuid.uuid4().hex] = {
+            "time_sent": now,
+            "rel_channel": router_data["rel_channel"],
+            "router_token": router_token,
+            "payload": payload}
         self.ap_settings.metrics.increment(
-            "updates.client.bridge.apns.succeed",
+            "updates.client.bridge.apns.%s.attempted" %
+            router_data["rel_channel"],
             self._base_tags)
+
+        apns_client.gateway_server.send_notification(router_token, payload,
+                                                     now)
         location = "%s/m/%s" % (self.ap_settings.endpoint_url,
                                 notification.version)
+        self.ap_settings.metrics.increment(
+            "updates.client.bridge.apns.%s.succeed" %
+            router_data["rel_channel"],
+            self._base_tags)
         return RouterResponse(status_code=201, response_body="",
                               headers={"TTL": notification.ttl,
                                        "Location": location},
                               logged_status=200)
 
+        # cleanup sent messages
+    def _cleanup(self, response):
+        now = time.time()
+        keys = self.messages.keys()
+        for msg_id in keys:
+            message = self.messages[msg_id]
+            expry = self._config[message['rel_channel']].get("expry", 10)
+            if message["time_sent"] < now - expry:
+                del self.messages[msg_id]
+        return response
+
     def _error(self, err):
         """Error handler"""
         if err['status'] == 0:
```

Comment:
Nit: `self.messages.pop(err['identifier'], None)` might be better; `del` will throw a `KeyError` if the entry was already removed from the dict.